### ISSUE TITLE
Memory issues with new 0.2 release

### ISSUE THREAD

**Issue by emauryg** [mlverse/torch#396]
I have been trying to re-run code that I have previously run successfully with the previous version but now I get a memory error even when using the same Windows PC as before, any suggestion at what might be the problem? 

`Error in (function (self, other)  : 
  [enforce fail at ..\..\c10\core\CPUAllocator.cpp:73] data. DefaultCPUAllocator: not enough memory: you tried to allocate 35389440 bytes. Buy new RAM!
00007FFAE022ADD800007FFAE022AC40 c10.dll!c10::MessageLogger::~MessageLogger [<unknown file> @ <unknown line number>]
00007FFAE022B2C300007FFAE022B290 c10.dll!c10::ThrowEnforceNotMet [<unknown file> @ <unknown line number>]
00007FFAE020BAFD00007FFAE020B990 c10.dll!c10::alloc_cpu [<unknown file> @ <unknown line number>]
00007FFAE020BBD200007FFAE020BBB0 c10.dll!c10::DefaultCPUAllocator::allocate [<unknown file> @ <unknown line number>]
00007FFA8655586200007FFA865556E0 torch_cpu.dll!at::native::empty_cpu [<unknown file> @ <unknown line number>]
00007FFA8675DB7F00007FFA866C3910 torch_cpu.dll!at::native::mkldnn_sigmoid_ [<unknown file> @ <unknown line number>]
00007FFA866EC10500007FFA866C3910 torch_cpu.dll!at::native::mkldnn_sigmoid_ [<unknown file> @ <unknown line number>]
00007FFA866C8D6200007FFA866C3910 torch_cpu.dll!at::native::mk `

**Comment by dfalbel** [comment]
I don't think there was any major change that could cause this. Would you share the code you are running so I can investigate?
Do you remember if before the update it was almost reaching the memory limits?

**Comment by emauryg** [comment]
The code is rather complex but the error occurs when I try to run the following for loop while when I run it once it works fine. In the past while it was a computationally intensive procedure, I've never found this issue (even when the code was less efficient). 

`errors = rep(NA,K+3)
elbos = rep(NA, K+3)
dC = torch_diag_embed(count_matrix$sum(c(1,2))) ## total number of mutations per sample
for (k in 3:(K+3)){
  init_pars <- init_nmf(count_matrix = count_matrix,K=k,G=G,P = P,S = S,Q = Q,X = xmat)
  mod1 = runEM(init_pars, count_matrix, k, G, P, S, Q, X=xmat)
  T_tensor = mod1$T_tensor
  theta = mod1$theta
  EC = T_tensor$matmul(theta)$matmul(dC)
  A = count_matrix - EC
  err_dist = 2* (count_matrix*(torch_log(count_matrix + 1e-14) - torch_log(EC + 1e-14)) - A)$sum()
  cat("============================== \n")
  cat("K = ",k,"; Error distance: ", as_array(err_dist),"\n")
  cat("=============================== \n")
  errors[k] = as_array(err_dist)
  elbos[k] = as_array(mod1$elbo)
   
}`

**Comment by dfalbel** [comment]
It's hard to tell what could be causing.. Maybe you can take a look at the full list of changes for this release and see if something rings the bell: https://github.com/mlverse/torch/blob/master/NEWS.md

Maybe that could be a change in LibTorch (we updated to libtorch 1.7 in this release) and it's making it more memory hungry.

You could also try adding a `gc()` call in the loop and monitor the memory consumption.

**Comment by emauryg** [comment]
Oh yeah, it seems that out of nowhere the code requests more than the maximum stated in gc(). Again, this was not an issue previously.

**Comment by mattwarkentin** [comment]
I have been getting a similar "not enough memory" error, not sure why...

**Comment by dfalbel** [comment]
I am investigating that issue but still not sure what could be causing it. One thing that is known is that, if you run this code without the `gc()` call, it will eventually get an OOM error. 

```
x <- torch::torch_randn(10000, 10000)
for (i in 1:100) {
  x <- torch::torch_relu(x)
  gc()
}
```

This is because the R garbage collector is lazy and will only run when R has to deallocate memory, but R is not aware of the memory allocated by libtorch so it doesn't think that it needs to cleanup the references to the intermediate references. This is very similar to this issue in arrow: https://issues.apache.org/jira/browse/ARROW-10080

But this is not new in the 0.2.0 release, so I wonder what could be causing the new behavior.

**Comment by emauryg** [comment]
With the gc() it runs ok, but then if I try to re run the code, I get an OOM error. I basically need to restart my Rstudio session to make the for loop run smoothly. Is there a way to ensure that the loop does a sufficient number of allocations?

**Comment by dfalbel** [comment]
Hi @emauryg ! I have fixed the `gc` handling in the dev version. Could you test it and see id the problem is fixed?
You can install the dev version with:

```
remotes::install_github("mlverse/torch")
```

**Comment by emauryg** [comment]
Just tested it and it seems that now using the gc() at the end of the for loop fixes the issue! Thank you for addressing this so quickly.

**Comment by dfalbel** [comment]
Thank you ! Actually, you shouldn't need the `gc` anymore too. Let me know if that works!

**Comment by mattwarkentin** [comment]
Do you call `gc()` internally now, @dfalbel?

**Comment by emauryg** [comment]
Ran it without the gc() now and it seems to work as well.

**Comment by dfalbel** [comment]
Yes! I will call gc internally in 2 occasions: 

1. The CPU allocator failed to allocate more memory. In this case we call gc and retry.

2. After each 4GB of memory allocated in the torch side. There's an option to configure it. (torch.threshold_call_gc).

I believe this will fix most issues with the gc laziness.

**Comment by emauryg** [comment]
I seem to be running a similar issue again, even if calling gc() after the loops. Here is just a simple optimization problem, not sure why torch would try to allocate more memory rather than just replacing values:

`optimize_TS <- function(settings, count_matrix, learning_rate=0.1, K, G, S, P){
  old_loss = 1e11
  theta <-  nnf_softmax(torch_cat(c(settings$eta, torch_zeros(1,S)),dim=1), dim=1)
  settings[["K"]] = K
  settings[["G"]] = G
  settings[["S"]] = S
  settings[["P"]] = P
  settings[["count_matrix"]] = count_matrix
  settings[["dmat"]] = dmat
  MODEL = estimate_TS(theta_val = theta, count_matrix = count_matrix, settings = settings)
  optimizer = optim_adam(MODEL$parameters, lr = learning_rate)
  it=0
  not_converged =TRUE
  while (not_converged == TRUE && it != 50000){
    it = it+1
    TS_est = MODEL()
    optimizer$zero_grad()
    new_loss = TS_est
    new_loss$backward()
    optimizer$step()
    
    if (it %% 100 == 0 || it==1){
      cat("TS Step:",it,":", as.numeric(new_loss$item()), "\n")
    }
    
    if(abs(new_loss$item() - old_loss)/old_loss < 1e-6){
      not_converged = FALSE
    } else {
      old_loss = new_loss$item()
    }
    gc()
  }
  
  return(get_pars_TS(MODEL))
}`

**Comment by dfalbel** [comment]
Do you get an error message or the session just hangs?

**Comment by emauryg** [comment]
Last time I got the error message, and then the session kind of crashed.
